Optimize OH by hemisphere instead of globally

msulprizio opened the pull request:
### Name and Institution (Required)

Name: Melissa Sulprizio
Institution: Harvard

### Describe the update

OH is now optimized by hemisphere instead of globally. This is done by creating two run directories in `jacobian_runs` instead of one for the OH perturbation. Both simulations utilize a new netCDF mask file `ExtData/HEMCO/MASKS/v2024-08/hemisphere_mask.01x01.nc` where the Northern Hemisphere grid cells are assigned ID=1 and Southern Hemisphere grid cells are assigned ID=2. The code in `jacobian.sh` automatically updates `HEMCO_Config.rc` to modify/add the following lines:

```
2 OH_pert_factor Perturbations_OH.txt - - - xy 1 1

* HEMIS_MASK $ROOT/MASKS/v2024-08/hemisphere_mask.01x01.nc Hemisphere 2000/1/1/0 C xy 1 * - 1 1 
```

It also creates a new file `Perturbations_OH.txt` with contents:

```
# Input file for applying emissions perturbations to state vector elements via HEMCO

# Mask field name
HEMIS_MASK

# Name   | ID  | Scale factor
DEFAULT    0     1.0
N_HEMIS    1     1.1
```

The same is done for the second run directory for the Southern Hemispheric perturbation (using ID=2 in Perturbations_OH.txt).

laestrada reviewed (changes requested):
Thanks Melissa, comments inline!

Also, is the `ExtData/HEMCO/MASKS/v2024-08/hemisphere_mask.01x01.nc` file on S3?

msulprizio commented on src/components/jacobian_component/jacobian.sh:
```diff
@@ -265,18 +297,19 @@ create_simulation_dir() {
     # as a CH4 tracer in the configuraton files
     if is_number "$x"; then
-        if [ $x -gt 0 ] && [ "$BC_elem" = false ] && [ "$OH_elem" = false ]; then
+        if [ $x -gt 0 ]; then
```
This update includes adding new tracers for the BC and OH perturbation simulations for consistency with the emissions perturbation simulations.

msulprizio replied:
I have reverted this code in https://github.com/geoschem/integrated_methane_inversion/pull/266/commits/9e93e2637b556b49ba037906d1a636fa666040d7

msulprizio commented on src/components/kalman_component/kalman.sh:
```diff
@@ -38,7 +38,7 @@ setup_kf() {
     fi
     if "$OptimizeOH";then
-	nElements=$((nElements+1))
+	nElements=$((nElements+2))
```
Yes, good catch

msulprizio replied:
Now fixes in https://github.com/geoschem/integrated_methane_inversion/pull/266/commits/82b1994665329f26ab4e8028c227ec627449608a

msulprizio commented on src/inversion_scripts/operators/operator_utilities.py:
```diff
@@ -171,34 +182,17 @@ def concat_tracers(run_id, gc_date, config, sv_elems, n_elements, baserun=False)
         chunks = 'auto'
     )
     keepvars = [f'SpeciesConcVV_CH4_{i:04}' for i in sv_elems]
 
```
As mentioned above, this PR includes changing BC and OH jacobian rundirs to include an additional tracer now named for the current state vector element. This was done for consistency with how the emission perturbations are handled and to avoid the need for different treatment of tracer names throughout IMI scripts.

msulprizio commented on src/components/statevector_component/statevector.sh:
```diff
@@ -105,7 +105,7 @@ reduce_dimension() {
     fi
     if "$OptimizeOH";then
-	nElements=$((nElements+1))
+	nElements=$((nElements+2))
```
Now fixed in https://github.com/geoschem/integrated_methane_inversion/pull/266/commits/82b1994665329f26ab4e8028c227ec627449608a

msulprizio commented on src/components/setup_component/setup.sh:
```diff
@@ -141,7 +141,7 @@ setup_imi() {
     fi
     if "$OptimizeOH";then
-	nElements=$((nElements+1))
+	nElements=$((nElements+2))
```
Now fixed in https://github.com/geoschem/integrated_methane_inversion/pull/266/commits/82b1994665329f26ab4e8028c227ec627449608a

msulprizio commented:
Thanks for reviewing @laestrada. I think I have addressed the bugs you found. Let me know if you want to discuss the changing of CH4 tracer names to include element numbers for the BC and OH perturbation runs. I have no strong preferences either way.

Also, I just copied `ExtData/HEMCO/MASKS/v2024-08/hemisphere_mask.01x01.nc` to S3. see https://s3.amazonaws.com/gcgrid/index.html#HEMCO/MASKS/v2024-08/.

msulprizio commented on src/components/jacobian_component/jacobian.sh:
```diff
@@ -185,11 +192,13 @@ create_simulation_dir() {
+    if [[ $x -eq 0 ]]; then
+	sed -i -e "s:#EmisCH4_Total:EmisCH4_Total:g" HEMCO_Diagn.rc
+	sed -i -e "s:#EmisCH4_SoilAbsorb:EmisCH4_SoilAbsorb:g" HEMCO_Diagn.rc
+    fi
```
Now removed in 82b1994

laestrada reviewed (approved):
Looks good, thanks Melissa!

megan-he commented on src/inversion_scripts/operators/operator_utilities.py:
In lines 198-201 `check_is_OH_element` is missing the `is_regional` argument

megan-he reviewed (commented):
Also, I believe in `src/inversion_scripts/invert.py`:

- [this](https://github.com/geoschem/integrated_methane_inversion/blob/eb3cfce59b526c819ab6e8bddc447865c2237a18/src/inversion_scripts/invert.py#L227) line should be changed to
`OH_weight = 2/(n_elements-2)` to reflect the 2 hemispheric OH elements
- [these](https://github.com/geoschem/integrated_methane_inversion/blob/eb3cfce59b526c819ab6e8bddc447865c2237a18/src/inversion_scripts/invert.py#L254C5-L256C40) lines should be:
```
if oh_optimization:
        xhat[-2:] += 1
        print(f"xhat[OH] for Northern Hemisphere = {xhat[-2]}")
        print(f"xhat[OH] for Southern Hemisphere = {xhat[-1]}")
```

megan-he commented:
I think both OH perturbation factors still need to be correctly applied to the posterior simulation in `HEMCO_Config.rc `.

msulprizio commented:
Thanks for the comments @megan-he. I have now addressed your comments in commit https://github.com/geoschem/integrated_methane_inversion/pull/266/commits/204124c2a321a2c568ae76448d3075c7f4bc6484.

Also tagging @laestrada

megan-he reviewed (approved):
Tested this and it ran through. Thank you Melissa!
